src/pages/experiments/[experimentId]/data-exploration/components/heatmap/HeatmapMetadataTrackSettings.jsx

```diff
@@ -18,23 +18,58 @@ const HeatmapMetadataTrackSettings = () => {
   const cellSets = useSelector((state) => state.cellSets);
   const selectedTracks = useSelector((state) => state.componentConfig.interactiveHeatmap.config.selectedTracks);
 
-  const [trackData, setTrackData] = useState([]);
+  const getCellSets = (category) => {
+    if (!cellSets || cellSets.loading) {
+      return [];
+    }
 
-  useEffect(() => {
-    setTrackData(
-      _.unionBy(
-        trackData,
-        getCellSets(
-          ['cellSets', 'metadataCategorical'],
-        ).map(
-          (data) => ({ selected: selectedTracks.includes(data.key), key: data.key }),
-        ),
-        'key',
+    return cellSets.hierarchy.map(
+      ({ key }) => (
+        { key, name: cellSets.properties[key].name, type: cellSets.properties[key].type }
       ),
+    ).filter(
+      ({ type }) => category.includes(type),
     );
+  };
+
+  const getInitialTrackData = () => {
+    return getCellSets(
+      ['cellSets', 'metadataCategorical'],
+    ).map(
+      (data) => ({ selected: selectedTracks.includes(data.key), key: data.key }),
+    );
+  }
+
+  const isInitialRenderRef = useRef(true);
+  const [trackData, setTrackData] = useState(getInitialTrackData());
+
+  const getUpdatedTrackData = () => {
+    return _.unionBy(
+      trackData,
+      getCellSets(
+        ['cellSets', 'metadataCategorical'],
+      ).map(
+        (data) => ({ selected: selectedTracks.includes(data.key), key: data.key }),
+      ),
+      'key',
+    )
+  };
```

Comment:
Looking at `getInitialTrackData` and the `getUpdatedTrackData` functions, they are almost the same apart from the union and `trackData` parameter. You could have only one function for getting trackData, which will be with the same contents as `getUpdatedTrackData` at the moment, and restore the initial default for trackData which was an empty list.

Reply:
I am going to refactor the reused code, but about the initial default it is actually a bit of a problem:

The reason I am setting the original default as `getInitialTrackData` is that before the way this menu was filled before was also triggering the first update to the heatmap here:
```

  useEffect(() => {
    // Prevent initial dispatch when object appears
    if (isInitialRenderRef.current) {
      return;
    }

    setTrackData(getUpdatedTrackData());
  }, [cellSets.hierarchy]);

```

Since now I am preventing this initial `setTrackData` the menu is not being filled, so I need to fill it by default.